cc-wrapper: don't use system rm and mktemp

IlyaNiklyaev opened the pull request:
## Description of changes

When response file is enabled (which is default for clang), cc-wrapper uses `rm` and `mktemp` commands from external `$PATH`. It breaks clang wrapper on systems w/o such tools, and using system tools is not a good idea in general. So we just take these tools from coretools explicitly.

On `master`:
```
$ PATH= $(nix build .#clang_16 --print-out-paths)/bin/c++
/nix/store/1n3cfg8hnh6izkjfnw0xwdz7h7wszyv4-clang-wrapper-16.0.6/bin/c++: line 249: mktemp: No such file or directory
```

On this PR:
```
$ PATH= $(nix build .#clang_16 --print-out-paths)/bin/c++
clang-16: error: no input files
```

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

matshch commented:
@ofborg build tests.cc-wrapper

sdht0 commented on pkgs/build-support/cc-wrapper/cc-wrapper.sh:
```diff
@@ -246,8 +247,8 @@ if [[ -e @out@/nix-support/cc-wrapper-hook ]]; then
-    responseFile=$(mktemp "${TMPDIR:-/tmp}/cc-params.XXXXXX")
-    trap 'rm -f -- "$responseFile"' EXIT
+    responseFile=$(PATH=$path_utils mktemp "${TMPDIR:-/tmp}/cc-params.XXXXXX")
+    trap 'PATH=$path_utils rm -f -- "$responseFile"' EXIT
```
Why does PATH need to be explicitly set here again when it is already set in line 15/16?

matshch replied:
It is being restored to its original value in line 240.

IlyaNiklyaev replied:
Yep, that's why this PR is needed. PATH restoration on line 240 is done for the hook a couple lines after. We could place PATH restoration after the hook and this response file-related code, but I'm not sure it won't break hooks somewhere.

sdht0 replied:
Ah right. Maybe we move that line inside the if and else instead of creating another variable?

sdht0 replied:
Although `cc-wrapper-hook` probably relies on the original PATH. So nvm.

IlyaNiklyaev replied:
I think I found a better solution: instead of playing with `$PATH` variable we may just call `rm` and `mktemp` via absolute paths like `@coreutils_bin@/bin/mktemp`. Also it makes our intent to not use system paths much clearer. I'll try this and return :)

IlyaNiklyaev replied:
Yep, that worked.

matshch commented:
It is pretty usual for ofborg to timeout on stdenv rebuilds, at least we have tried. I'll try running these tests myself, starting with aarch64-darwin, but it will take some time.

IlyaNiklyaev commented:
> I'll try running these tests myself, starting with aarch64-darwin

Let's share this work :) I could run some tests tomorrow.

matshch commented:
I've got the same undefined reference errors on aarch64-linux on nixos-unstable with and without this PR, looks like #285016.

IlyaNiklyaev commented:
> I've got the same undefined reference errors on aarch64-linux on nixos-unstable with and without this PR, looks like #285016.

Oops, I mean `aarch64-linux` here, not `aarch64-darwin` :( https://github.com/NixOS/nixpkgs/pull/310195#issuecomment-2105215370

IlyaNiklyaev committed:
cc-wrapper: don't use system rm and mktemp

When response file is enabled (which is default for clang), cc-wrapper uses
rm and mktemp commands from external $PATH. It breaks clang wrapper on
systems w/o such tools, and using system tools it's not a good idea in general.
So we just take these tools from the nix coretools.

IlyaNiklyaev commented:
I think it looks good now. @Ericson2314 could you please take a look on this PR?

matshch reviewed (approved):
This patch seems to work, and `tests.cc-wrapper` builds successfully on x86_64-linux.

matshch commented:
On aarch64-linux I have successfully built:
* tests.cc-wrapper.default
* tests.cc-wrapper.llvmTests (it has recurseIntoAttrs)
* tests.cc-wrapper.gccTests.gccStdenv
* tests.cc-wrapper.gccTests.gcc8Stdenv
* tests.cc-wrapper.gccTests.gcc9Stdenv
* tests.cc-wrapper.gccTests.gcc10Stdenv
* tests.cc-wrapper.gccTests.gcc11Stdenv
* tests.cc-wrapper.gccTests.gcc12Stdenv
* tests.cc-wrapper.gccTests.gcc13Stdenv

Next attributes are broken in nixos-unstable, nixos-23.11 and in this PR:
* tests.cc-wrapper.gccTests.gcc49Stdenv (it has some other issue about atomics, probably too old for them)
* tests.cc-wrapper.gccTests.gcc6Stdenv (due to #285016)
* tests.cc-wrapper.gccTests.gcc7Stdenv (due to #285016)
* tests.cc-wrapper.supported (it depends on both attributes above)

matshch commented:
tests.cc-wrapper builds just fine on aarch64-darwin and x86_64-darwin.

cormacrelf commented:
Sorry, I missed this PR when I filed #316674. The fix in #316700 seems better to me because `@coreutils_bin@` is sometimes empty, when the `nativeTools` flag in `cc-wrapper/default.nix` is true. In which case this approach ends up as `/bin/mktemp`. Which is not quite correct, but will limp on long enough that someone will rediscover this issue in 5 years.

IlyaNiklyaev commented:
@cormacrelf yep, that's unfortunate, I'll try to fix that. The PR you mentioned has it's own issue, as I can see https://github.com/NixOS/nixpkgs/pull/316700/files#r1623952918

Artturin commented:
https://github.com/NixOS/nixpkgs/pull/349589 (Added you as a co-author in the commit)

https://github.com/NixOS/nixpkgs/pull/349585
